File: frontend/src/objects/customer/profile/Coupon.ts
// Business note: frontend mirror of a backend service object; keep names and fields aligned unless this file is explicitly frontend-only.
import type {
  CouponId,
  CurrencyCents,
  DescriptionText,
  DisplayText,
  IsoDateTime,
} from '@/objects/core/SharedObjects'

export type Coupon = {
  id: CouponId
  title: DisplayText
  discountCents: CurrencyCents
  minimumSpendCents: CurrencyCents
  description: DescriptionText
  expiresAt: IsoDateTime
}

File: backend/src/main/scala/services/customer/objects/Coupon.scala
package services.customer.objects

// Business note: service-owned domain object; mirror it in the frontend only when it is part of the protocol surface.
import system.objects.given

import io.circe.{Decoder, Encoder}
import io.circe.generic.semiauto.*
import system.objects.*

final case class Coupon(
    id: CouponId,
    title: DisplayText,
    discountCents: CurrencyCents,
    minimumSpendCents: CurrencyCents,
    description: DescriptionText,
    expiresAt: IsoDateTime,
)
object Coupon:
  given Encoder[Coupon] = deriveEncoder
  given Decoder[Coupon] = deriveDecoder
